### PR TITLE
Extract method for rendering raw DOT-encoded graphs

### PR DESCRIPTION
**Pull request by fwcd**
This lets users render raw DOT-encoded graphs using the library. Despite maybe being an uncommon use case, it entails few code changes apart from making the renderer slightly more modular.

### CODE REVIEW

File: Sources/DOT/Graph+Rendering.swift [hunk 1]
```diff
@@ -29,10 +29,8 @@ fileprivate func which(_ command: String) throws -> URL {
     return URL(fileURLWithPath: string)
 }
 
-extension Graph {
-    public func render(using layout: LayoutAlgorithm, to format: Format) throws -> Data {
-        let encoded = DOTEncoder().encode(self)
-
+extension String {
+    public func renderAsDOT(using layout: LayoutAlgorithm, to format: Format) throws -> Data {
```
Comment (mattt) [inline Sources/DOT/Graph+Rendering.swift]: `String` is a commonly used type in Swift, and adding a GraphViz-specific convenience method in an extension has the effect of polluting the API surface area with unrelated functionality.

If the intention is to decouple graph rendering to allow for raw DOT to be rendered by `dot`, then I think a better design would be to create a `Renderer` type that can process both `Graph` and `String` values. The advantage there is that a `Renderer` type could be responsible for formats and layout algorithms, which is currently delegated to `Graph` in a way that arguably conflates model and view concerns.

**Review by mattt — CHANGES_REQUESTED** [review]
Thanks for this PR, @fwcd! I agree that this would be useful to have, and have a slightly different idea for how to approach this.

File: Sources/DOT/Graph+Rendering.swift (hunk 1 above)
Comment (fwcd) [reply Sources/DOT/Graph+Rendering.swift]: Fixed!

File: Sources/DOT/DOTRenderer.swift [hunk 2]
```diff
@@ -29,10 +29,16 @@ fileprivate func which(_ command: String) throws -> URL {
     return URL(fileURLWithPath: string)
 }
 
-extension Graph {
-    public func render(using layout: LayoutAlgorithm, to format: Format) throws -> Data {
-        let encoded = DOTEncoder().encode(self)
+public struct DOTRenderer {
+    private let layout: LayoutAlgorithm
```
Comment (mattt) [inline Sources/DOT/DOTRenderer.swift]:
```suggestion
    let layout: LayoutAlgorithm
```

No harm in letting API consumers be able to tell which layout a renderer uses. I could even see a case for making this a `var`.

File: Sources/DOT/DOTRenderer.swift [hunk 3]
```diff
@@ -29,10 +29,16 @@ fileprivate func which(_ command: String) throws -> URL {
...
+    private let format: Format
 
+    public init(using layout: LayoutAlgorithm, to format: Format) {
+        self.layout = layout
+        self.format = format
+    }
+
+    public func render(dotEncoded: String) throws -> Data {
```
Comment (mattt) [inline Sources/DOT/DOTRenderer.swift]:
```suggestion
    public init(using layout: LayoutAlgorithm) {
        self.layout = layout
    }

    public func render(dotEncoded: String, to format: Format) throws -> Data {
```

**Review by mattt — COMMENTED** [review]
I just realized that I left my review pending this whole time. Submitting this now, in spite of the changes made in the meantime. (They may very well be outdated now).

**Comment by mattt** [comment]
Thanks again for your work on this and your patience. I went ahead and made the changes I wanted for this feature. These will be available in the forthcoming 0.2.0 release. Let me know if you have any questions or feedback on these. Thanks!

**Pull request merged**